### DISCUSSION TITLE
Define different workflow on different branches with workflow_dispatch

### DISCUSSION THREAD

**Discussion by caiofaustino** [community/community#25412] [Actions]
<p>Hi there folks,</p>
<p>What I’m trying to achieve is to have 2 different protected branches with different workflows.</p>
<p>The first one has workflow <code>foo</code> that runs <code>on: pull_request: branches: [ branchA ]</code><br>
I want to have a similar workflow on <code>branchB</code> and I wouldn’t mind calling it <code>foo</code> as well since they perform the same function, but I understand this could be confusing so I called it <code>bar</code>.</p>
<p>The thing is, I’m adding the workflow file only to the branch it applies to, so <code>branchA</code> has the file for workflow <code>foo</code> and <code>branchB</code> has the file for workflow <code>bar</code>. I think this is the correct approach since I only want each workflow to run on that specific branch. Is this a fair assumption?</p>
<p>Also I want to trigger workflow <code>bar</code> manually so I only set <code>on:  workflow_dispatch:</code>. But then after merging it I can’t see the action the the GitHub UI. Is there something else I need to do to add the Action properly other than commit the file? When testing <code>bar</code> with a PR trigger it runs fine.</p>
<p>Thanks in advance</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/caiofaustino">@caiofaustino</a>,</p>
<p>About your first question,</p>
<blockquote>
<p>I think this is the correct approach since I only want each workflow to run on that specific branch. Is this a fair assumption?</p>
</blockquote>
<p>Sure, you can set up and run different workflows on different branches.<br>
However, you also need to know that some events only trigger workflow runs if the workflow files are on the default branch, such as the events <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#repository_dispatch" rel="noopener nofollow ugc"><strong>repository_dispatch</strong></a>, <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#check_run" rel="noopener nofollow ugc"><strong>check_run</strong></a>, <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#check_suite" rel="noopener nofollow ugc"><strong>check_suite</strong></a>, etc. . These events are not available on other branches.</p>
<br>
<p>About your second question,</p>
<blockquote>
<p>Also I want to trigger workflow  <code>bar</code>  manually so I only set  <code>on: workflow_dispatch:</code> . But then after merging it I can’t see the action the the GitHub UI. Is there something else I need to do to add the Action properly other than commit the file?</p>
</blockquote>
<p>Normally, when the workflow file is not on the on the default branch, the workflow will not be listed under <strong>All workflows</strong> on the <strong>Actions</strong> tab, unless the workflow has been triggered at least once.</p>
<p>For the <strong>workflow_dispatch</strong> event, if you want to manually trigger the <strong>workflow_dispatch</strong> event on GitHub UI, the workflow file must be also on the default branch. If the workflow file only on other branches and not on the default branch, you need to use the “<a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-workflow-dispatch-event" rel="noopener nofollow ugc"><strong>Create a workflow dispatch event</strong></a>” API to trigger the <strong>workflow_dispatch</strong> event on these branches. For more details, you can see “<a href="https://docs.github.com/en/free-pro-team@latest/actions/managing-workflow-runs/manually-running-a-workflow" rel="noopener nofollow ugc"><strong>Manually running a workflow</strong></a>”.</p>

**Comment by caiofaustino** [comment]
<p>Thanks a lot! I’ll give this a try and come back to accept your answer as soon as I have it working :slight_smile:</p>

**Comment by miguelaferreira** [comment]
<p>Hi!</p>
<p>According to the <a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-workflow-dispatch-event" rel="noopener nofollow ugc">docs for creating a workflow dispatch event</a> the payload of the POST has  a <code>ref</code> attribute that should define the source branch/tag for the workflow. However, that <code>ref</code> only needs to point to an existing branch, and the workflow will only be dispatched if it is present in the default branch.</p>
<p>I’ve created a test repo with a PR that shows this behaviour <a href="https://github.com/miguelaferreira/test-gha-workflow-dispatch/pull/1" rel="noopener nofollow ugc">https://github.com/miguelaferreira/test-gha-workflow-dispatch/pull/1</a></p>
<p>The docs and the implementation don’t seem to align here.</p>
